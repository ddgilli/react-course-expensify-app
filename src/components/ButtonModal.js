import React from 'react';
import Modal from 'react-modal';

const ButtonModal = (props) => ( //implicit return statement
  <Modal
    isOpen={!!props.idToRemove} //required
    closeTimeoutMS={200}
    contentLabel = {"Confirm Delete?"} //required
    className="modal"
    onRequestClose ={props.onCancel}
  >
    <h3 className="modal__title">Confirm Delete?</h3>
    <button className="button"onClick={props.removeExpense}>Yes</button>
    <button className="button"onClick={props.onCancel}>Cancel</button>
  </Modal>
);

export default ButtonModal;