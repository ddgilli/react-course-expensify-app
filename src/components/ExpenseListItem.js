import React from 'react';
import { connect} from 'react-redux';
import { Link } from 'react-router-dom';

const ExpenseListItem = ({ id, description, amount, createdAt}) => (
<div>
  <h3><Link to={`edit/${id}`}>Description: {description}</Link></h3>
  <p>Amount: {amount} - CreatedAt: {createdAt}</p>
</div>
);

export default ExpenseListItem;

