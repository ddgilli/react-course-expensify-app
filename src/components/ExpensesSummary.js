import React from 'react';
import { connect} from 'react-redux';
import numeral from 'numeral';
import selectExpenses from '../selectors/expenses';
import selectExpensesTotal from '../selectors/expenses-total';

export const ExpensesSummary = ({ expenseCount, expensesTotal }) => {
  const expenseWord = expenseCount === 1 ? 'expense ' : 'expenses ';
  const formattedTotal = numeral(expensesTotal / 100).format('$0,0.00');
  return (

   <div>  
    <h1>Viewing {expenseCount} {expenseWord} with a total of {formattedTotal}</h1>
  </div>
   
  )};

//this defines the state fields (already in the store - see configureStore.js
//for specifics) that we want from the store
const mapStateToProps = (state) => {
   {
    const visibleExpenses = selectExpenses(state.expenses, state.filters);
    return {
      expenseCount: visibleExpenses.length,
      expensesTotal: selectExpensesTotal(visibleExpenses)
    }
  };
};

//this actually performs the mapping - the objects in mapStateToProps
//tell the connect function what we want from the store, the second
//argument is the return component - ExpensesSummary - which we just created.
export default connect(mapStateToProps)(ExpensesSummary);