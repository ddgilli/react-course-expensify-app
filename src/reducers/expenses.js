const expensesReducerDefaultState = [];//defaults can be much more verbose

export default (state = expensesReducerDefaultState, action) => {
  switch(action.type) {
    case 'ADD_EXPENSE':
      return [
        ...state,
        action.expense
      ];
    case 'REMOVE_EXPENSE':
    return state.filter(({ id }) => id !== action.id);
//old way    return state.filter((expense) => action.expense.id !== expense.id)
    case 'EDIT_EXPENSE':
      //retrieve expense using id, update field with new value 
      return state.map((expense) => {
        if (expense.id === action.id)
        {
          return {
            ...expense, 
            ...action.updates
          };
         } else {
          return expense;
        };
      });
    default:
      return state;
  }
};

