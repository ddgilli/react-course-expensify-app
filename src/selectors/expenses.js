//timestamps - (milliseconds)   1000 milli = 1 sec
//positive or negative integers
//January 1st 1970 is day zero - before is neg, after is pos
import moment from 'moment';

//Get visible expenses

export default (expenses, {text, sortBy, startDate, endDate}) => {
  return expenses.filter((expense) => {
    const createdAtMoment = moment(expense.createdAt);
    const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
    const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
    const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());
 
    return startDateMatch && endDateMatch && textMatch; //return if all are true
  }).sort((a,b) => {
    if (sortBy === 'date') {
      return a.createdAt < b.createdAt ? 1 : -1;
    } else if (sortBy === 'amount') {
        return a.amount < b.amount ? 1 : -1;
      }
    });
};