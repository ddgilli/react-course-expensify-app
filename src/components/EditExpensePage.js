import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { startEditExpense, startRemoveExpense } from '../actions/expenses';
import ButtonModal from './ButtonModal';

export class EditExpensePage extends React.Component {
  constructor (props) {
    super(props);
    this.state = {
      idToRemove: undefined
    }
  }

  onSubmit = (expense) => {
    this.props.startEditExpense(this.props.expense.id, expense);
    this.props.history.push('/');
  };

  onRemoveClick = () => {  
    this.setState(() => ({
      idToRemove: this.props.expense.id
    }));
  }

  removeExpense = () => {
    this.props.startRemoveExpense( {id: this.props.expense.id} );
    this.setState(() => ({
      idToRemove: undefined
      }));
    this.props.history.push('/');
  }

  onCancel = () => {
    this.setState(() => ({
      idToRemove: undefined
    }));
  }

  render() {
    return (
      <div>
        <div className="page-header">
          <div className="content-container"> 
            <h1 className="page-header__title">Edit Expense</h1>
          </div>
        </div>
        <div className="content-container">
          <ExpenseForm className="form" onSubmit={this.onSubmit} expense={this.props.expense} />
          <button className="button button--secondary" onClick={this.onRemoveClick}>Remove Expense</button>    
          <ButtonModal 
            className="modal"
            idToRemove={this.state.idToRemove} 
            onCancel={this.onCancel}
            removeExpense={this.removeExpense}
          />
          
        </div>
      </div>
    );
  }
};

const mapDispatchToProps = (dispatch, props) => ({    
  startRemoveExpense: ( data ) => dispatch(startRemoveExpense(data)),
  startEditExpense: (id, expense) => dispatch(startEditExpense(id, expense))
}); 


const mapStateToProps = (state, props) => {
  return {
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  };
};

export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);