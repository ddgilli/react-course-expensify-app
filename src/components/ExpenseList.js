import React from 'react';
import { connect} from 'react-redux';
import ExpenseListItem from './ExpenseListItem';
import selectExpenses from '../selectors/expenses';
import { removeExpense } from '../actions/expenses';

//local component starts with just a declaration
export const ExpenseList = (props) => (
  <div>
  {
    props.expenses.length === 0 ? (
      <p>There are no expenses, add some!</p>
    ) : (
      props.expenses.map((expense) => {
          return <ExpenseListItem key={expense.id} {...expense} />;
        })
    )
  }
      {}
  </div>
);

//this defines the state fields (already in the store - see configureStore.js
//for specifics) that we want from the store
const mapStateToProps = (state) => {
  return {
    expenses: selectExpenses(state.expenses, state.filters)
  };
};

//this actually performs the mapping - the objects in mapStateToProps
//tell the connect function what we want from the store, the second
//argument is the return component - ExpenseList - which we just created.
export default connect(mapStateToProps)(ExpenseList);

//     / mine
      // props.expenses.map((expense) => (
      //   <ExpenseListItem
      //      description={expense.description}
      //      amount={expense.amount}
      //      createdAt={expense.createdAt}
      //   />
      // ))

