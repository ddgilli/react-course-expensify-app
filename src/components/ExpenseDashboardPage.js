import React from 'react';
import ExpenseList from './ExpenseList';
import ExpensesSummary from './ExpensesSummary';
import ExpenseListFilters from './ExpenseListFilters';

const ExpenseDashboardPage = () => (
  <div>
    <ExpenseListFilters />
    <ExpenseList />
    <ExpensesSummary />
  </div>
);

export default ExpenseDashboardPage;
