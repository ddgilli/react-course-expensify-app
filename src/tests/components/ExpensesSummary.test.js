import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { ExpensesSummary } from '../../components/ExpensesSummary';


test('should render ExpensesSummary correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={1} expenseTotal={123.45}/>);
  expect(wrapper).toMatchSnapshot();
})

test('should reflect expenses length correctly', () => {
  const wrapper = shallow(<ExpensesSummary expenseCount={22} expenseTotal={123456789}/>);
  expect(wrapper).toMatchSnapshot();
})


//Viewing ${props.expenses.length} expenses with a total of $1