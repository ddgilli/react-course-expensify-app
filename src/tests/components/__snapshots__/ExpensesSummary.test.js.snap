// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should reflect expenses length correctly 1`] = `
<div>
  <h1>
    Viewing 
    22
     
    expenses 
     with a total of 
    $0.00
  </h1>
</div>
`;

exports[`should render ExpensesSummary correctly 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense 
     with a total of 
    $0.00
  </h1>
</div>
`;
